{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e19f0efa-6f24-436c-80cd-542186e0ec7d",
   "metadata": {},
   "source": [
    "Draft of scRNA-seq integration pipeline for joint analysis of GFP replicate samples 1-3. \n",
    "\n",
    "Resources: \n",
    "1) https://satijalab.org/seurat/articles/integration_introduction.html\n",
    "2) CART_I1.R from Zhaoyang"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "442cb4d7-2811-4b68-b355-2a75126e4ab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(data.table)\n",
    "library(tidyverse)\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "lapply(c(\"dplyr\",\"Seurat\",\"HGNChelper\",\"openxlsx\"), library, character.only = T)\n",
    "source(\"https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R\"); source(\"https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8ef5be2-9ea5-48cd-bfd4-927c3cd6d8d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data (pre-normalized and scaled to regress out percent-MT)\n",
    "# (load GFP, RBD, G1C separately)\n",
    "g1 <- readRDS(\"/oasis/tscc/scratch/mac008/seurat_objs/4RBD1_afterscale.rds\")\n",
    "g2 <- readRDS(\"/oasis/tscc/scratch/hsaha/jaesu-analysis-rds/5RBD2_afterscaling.rds\")\n",
    "g3 <- readRDS(\"/oasis/tscc/scratch/hsaha/jaesu-analysis-rds/6RBD3_afterscaling.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9905e39c-9f08-44ca-b845-133173b4fd41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify variable features for each dataset independently\n",
    "gfp.list <- c(g1,g2,g3)\n",
    "gfp.list <- lapply(X = gfp.list, FUN = function(x) {\n",
    "    x <- FindVariableFeatures(x, selection.method = \"mean.var.plot\", \n",
    "                              mean.cutoff = c(0.0125, 3), \n",
    "                              dispersion.cutoff = c(0.5, Inf))\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d14aacde-cb3c-4d82-a8f1-29402a694859",
   "metadata": {},
   "source": [
    "NOTE: CART_I1 analysis uses sctransformed data, skips above steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d107becf-7504-4359-bbb8-878624d50138",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select features that are repeatedly variable across datasets for integration\n",
    "features <- SelectIntegrationFeatures(object.list = gfp.list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2733f50e-0801-4c0d-abbd-20dd8584403a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gfp.anchors <- FindIntegrationAnchors(object.list = gfp.list, anchor.features = features)\n",
    "gfp.integrated <- IntegrateData(anchorset = gfp.anchors)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63ca1bce-b533-4a1c-ad93-6b4923fcc579",
   "metadata": {},
   "source": [
    "NOTE: CART_I1 analysis scales and runs pca on each dataset, then finds integration \n",
    "anchors by setting the reduction to \"rpca\". Additional parameters are set for creating the integrated object. I am currently using default parameters per the Seurat integration tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05a039dd-8cfd-43c6-904c-799b78d951eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(gfp.integrated, \"/oasis/tscc/scratch/mac008/seurat_objs/RBD4-6_afterintegrate.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bedf79e-7943-4bed-8f23-4fb1901c55b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gfp.integrated <- readRDS(\"/oasis/tscc/scratch/mac008/seurat_objs/RBD4-6_afterintegrate.rds\")\n",
    "\n",
    "# Run standard workflow\n",
    "gfp.integrated <- ScaleData(gfp.integrated, verbose = FALSE)\n",
    "\n",
    "saveRDS(gfp.integrated, \"/oasis/tscc/scratch/mac008/seurat_objs/RBD4-6_afterscale.rds\")\n",
    "\n",
    "gfp.integrated <- RunPCA(gfp.integrated, seed.use = 42)\n",
    "gfp.integrated <- RunTSNE(gfp.integrated, dims.use = 1:16, do.fast = T, perplexity =  15, seed.use = 7777)\n",
    "\n",
    "saveRDS(gfp.integrated, \"/oasis/tscc/scratch/mac008/seurat_objs/RBD4-6_aftertsne.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa4f062c-a460-48ea-8e6a-4c94100b67eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat <- readRDS(\"/oasis/tscc/scratch/mac008/seurat_objs/GFP1-3_aftertsne.rds\")\n",
    "\n",
    "DimPlot(seurat, reduction = \"tsne\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bba2f367-38bb-4134-83f8-2df6c9c4cc74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ----------------------- LABEL W/ NEW MARKER GENES -----------------------\n",
    "\n",
    "seurat <- readRDS(\"/oasis/tscc/scratch/mac008/seurat_objs/G1C10-12_aftertsne.rds\")\n",
    "\n",
    "seurat <- FindNeighbors(seurat, dims = 1:35)\n",
    "seurat <- FindClusters(seurat, resolution = 5, random.seed = 42)\n",
    "\n",
    "# Load in required files for SCType. SCType is used for labeling clusters.\n",
    "\n",
    "# load gene set preparation function\n",
    "source(\"https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R\")\n",
    "\n",
    "# load cell type annotation function\n",
    "source(\"https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R\")\n",
    "\n",
    "gs_list = gene_sets_prepare(\"https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_short.xlsx\", \"Immune system\")\n",
    "\n",
    "# Set the file path for db_ to the excel file for marker genes. This file should be formatted correctly. For guidance\n",
    "# on formatting, reference link in cell above in gs_list object.\n",
    "\n",
    "# DB file\n",
    "db_ = \"/home/mac008/scrna_project/nov15-W7F/jaesudata_markergenes_nov15.xlsx\";\n",
    "\n",
    "# Set the tissue type you are working with\n",
    "tissue = \"Bone marrow\" # e.g. Immune system,Pancreas,Liver,Eye,Kidney,Brain,Lung,Adrenal,Heart,Intestine,Muscle,Placenta,Spleen,Stomach,Thymus\n",
    "\n",
    "# Prepares the marker gene data for application to your results.\n",
    "gs_list = gene_sets_prepare(db_, tissue)\n",
    "\n",
    "# Performs calculation of distinguising cell types from marker genes. May take a minute or two.\n",
    "es.max = sctype_score(scRNAseqData = seurat[[\"integrated\"]]@scale.data, scaled = TRUE, \n",
    "                      gs = gs_list$gs_positive, gs2 = gs_list$gs_negative) \n",
    "\n",
    "# Changes the metadata of the seurat object to include the cell types you discovered in previous steps\n",
    "\n",
    "# NOTE: scRNAseqData parameter should correspond to your input scRNA-seq matrix. \n",
    "# In case Seurat is used, it is either pbmc[[\"RNA\"]]@scale.data (default), pbmc[[\"SCT\"]]@scale.data, in case sctransform is used for normalization,\n",
    "# or pbmc[[\"integrated\"]]@scale.data, in case a joint analysis of multiple single-cell datasets is performed.\n",
    "\n",
    "# merge by cluster\n",
    "cL_results = do.call(\"rbind\", lapply(unique(seurat@meta.data$seurat_clusters), function(cl){\n",
    "  es.max.cl = sort(rowSums(es.max[ ,rownames(seurat@meta.data[seurat@meta.data$seurat_clusters==cl, ])]), decreasing = !0)\n",
    "  head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(seurat@meta.data$seurat_clusters==cl)), 10)\n",
    "}))\n",
    "\n",
    "sctype_scores = cL_results %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  \n",
    "\n",
    "# Performs quality check by setting clusters that were not easily disguishable as \"unknown\"\n",
    "\n",
    "# set low-confident (low ScType score) clusters to \"unknown\"\n",
    "sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] = \"Unknown\"\n",
    "\n",
    "print(sctype_scores[,1:3])\n",
    "\n",
    "seurat@meta.data$customclassif = \"\"\n",
    "for(j in unique(sctype_scores$cluster)){\n",
    "  cl_type = sctype_scores[sctype_scores$cluster==j,]; \n",
    "  seurat@meta.data$customclassif[seurat@meta.data$seurat_clusters == j] = as.character(cl_type$type[1])\n",
    "}\n",
    "\n",
    "# Outputs final tSNE plot as pdf with clusters labeled\n",
    "# Change name based on resolution\n",
    "\n",
    "pdf(\"nov15-W7F/G1C10-12_markergenes_nov15.pdf\", width = 10, height = 10)\n",
    "plot <- DimPlot(seurat, reduction = \"tsne\", label = TRUE, label.size = 4, repel = TRUE, group.by = 'customclassif') \n",
    "print(plot+ ggtitle(\"G1C Integrated\"))\n",
    "dev.off()\n",
    "\n",
    "saveRDS(seurat, \"/oasis/tscc/scratch/mac008/seurat_objs/G1C10-12_afterlabels_markergenes_nov15.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b16402a-5cf9-42ef-bfa8-79065681a921",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ----------------------- OBTAIN CELL COUNTS PER CLUSTER -----------------------\n",
    "\n",
    "sample <- \"RBD4-6\"\n",
    "markergenes <- \"markergenes_nov15\"\n",
    "\n",
    "seurat <- readRDS(paste(\"/oasis/tscc/scratch/mac008/seurat_objs/\",sample,\"_afterlabels_\",markergenes,\".rds\",sep=\"\"))\n",
    "\n",
    "# Overwrite clusters from FindClusters with SC-Type labeled clusters in metadata\n",
    "Idents(seurat) <- \"customclassif\"\n",
    "\n",
    "md_df <- seurat@meta.data\n",
    "\n",
    "# Create a data.table from the metadata_df\n",
    "md_dt <- as.data.table(md_df)\n",
    "\n",
    "# Calculate the count of rows by orig.ident and customclassif\n",
    "md_count <- md_dt[, .N, by = c(\"orig.ident\", \"customclassif\")]\n",
    "\n",
    "# Reshape the data\n",
    "reshaped_table <- dcast(md_count, orig.ident ~ customclassif, value.var = \"N\")\n",
    "\n",
    "numeric_columns <- reshaped_table %>%\n",
    "  select_if(is.numeric)\n",
    "\n",
    "# Calculate row sums for each row in reshaped_table\n",
    "row_sums <- rowSums(numeric_columns <- reshaped_table %>%\n",
    "  select_if(is.numeric), na.rm = TRUE)\n",
    "\n",
    "# Add row sums as a new column to the table\n",
    "reshaped_table$TotalCells <- row_sums\n",
    "\n",
    "# Print or view the table\n",
    "reshaped_table\n",
    "\n",
    "write.csv(reshaped_table, \"nov15-W7F/rbd4-6_counts.csv\", row.names=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8132e5a1-6c08-45cc-afe2-c1030c3c0f13",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # Below is the pipeline using UMAP over tSNE (per Luca's request)\n",
    "# # Continues in place of 'run standard workflow' steps above\n",
    "# # ---------------------------------------------------------------\n",
    "\n",
    "# gfp.integrated <- readRDS(\"/oasis/tscc/scratch/mac008/seurat_objs/GFP1-3_afterscale.rds\")\n",
    "\n",
    "# gfp.integrated <- RunPCA(gfp.integrated, seed.use = 42)\n",
    "# gfp.integrated <- RunUMAP(gfp.integrated, dims = 1:16)\n",
    "\n",
    "# saveRDS(gfp.integrated, \"/oasis/tscc/scratch/mac008/seurat_objs/GFP1-3_afterumap.rds\")\n",
    "\n",
    "# DimPlot(gfp.integrated, reduction = \"umap\")\n",
    "\n",
    "# # ---------------------------------------------------------------\n",
    "\n",
    "# seurat <- readRDS(\"/oasis/tscc/scratch/mac008/seurat_objs/GFP1-3_afterumap.rds\")\n",
    "\n",
    "# seurat <- FindNeighbors(seurat, dims = 1:35)\n",
    "# seurat <- FindClusters(seurat, resolution = 5, random.seed = 42)\n",
    "\n",
    "# # Load in required files for SCType. SCType is used for labeling clusters.\n",
    "\n",
    "# # load gene set preparation function\n",
    "# source(\"https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R\")\n",
    "\n",
    "# # load cell type annotation function\n",
    "# source(\"https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R\")\n",
    "\n",
    "# gs_list = gene_sets_prepare(\"https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_short.xlsx\", \"Immune system\")\n",
    "\n",
    "# # Set the file path for db_ to the excel file for marker genes. This file should be formatted correctly. For guidance\n",
    "# # on formatting, reference link in cell above in gs_list object.\n",
    "\n",
    "# # DB file\n",
    "# db_ = \"/home/mac008/scrna_project/nov15-W7F/markergenes_nov15.xlsx\";\n",
    "\n",
    "# # Set the tissue type you are working with\n",
    "# tissue = \"Bone marrow\" # e.g. Immune system,Pancreas,Liver,Eye,Kidney,Brain,Lung,Adrenal,Heart,Intestine,Muscle,Placenta,Spleen,Stomach,Thymus\n",
    "\n",
    "# # Prepares the marker gene data for application to your results.\n",
    "# gs_list = gene_sets_prepare(db_, tissue)\n",
    "\n",
    "# # Performs calculation of distinguising cell types from marker genes. May take a minute or two.\n",
    "# es.max = sctype_score(scRNAseqData = seurat[[\"integrated\"]]@scale.data, scaled = TRUE, \n",
    "#                       gs = gs_list$gs_positive, gs2 = gs_list$gs_negative) \n",
    "\n",
    "# # Changes the metadata of the seurat object to include the cell types you discovered in previous steps\n",
    "\n",
    "# # NOTE: scRNAseqData parameter should correspond to your input scRNA-seq matrix. \n",
    "# # In case Seurat is used, it is either pbmc[[\"RNA\"]]@scale.data (default), pbmc[[\"SCT\"]]@scale.data, in case sctransform is used for normalization,\n",
    "# # or pbmc[[\"integrated\"]]@scale.data, in case a joint analysis of multiple single-cell datasets is performed.\n",
    "\n",
    "# # merge by cluster\n",
    "# cL_results = do.call(\"rbind\", lapply(unique(seurat@meta.data$seurat_clusters), function(cl){\n",
    "#   es.max.cl = sort(rowSums(es.max[ ,rownames(seurat@meta.data[seurat@meta.data$seurat_clusters==cl, ])]), decreasing = !0)\n",
    "#   head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(seurat@meta.data$seurat_clusters==cl)), 10)\n",
    "# }))\n",
    "\n",
    "# sctype_scores = cL_results %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  \n",
    "\n",
    "# # Performs quality check by setting clusters that were not easily disguishable as \"unknown\"\n",
    "\n",
    "# # set low-confident (low ScType score) clusters to \"unknown\"\n",
    "# sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] = \"Unknown\"\n",
    "\n",
    "# print(sctype_scores[,1:3])\n",
    "\n",
    "# seurat@meta.data$customclassif = \"\"\n",
    "# for(j in unique(sctype_scores$cluster)){\n",
    "#   cl_type = sctype_scores[sctype_scores$cluster==j,]; \n",
    "#   seurat@meta.data$customclassif[seurat@meta.data$seurat_clusters == j] = as.character(cl_type$type[1])\n",
    "# }\n",
    "\n",
    "# saveRDS(seurat, \"/oasis/tscc/scratch/mac008/seurat_objs/GFP_UMAP_afterlabels_markergenes_nov15.rds\")\n",
    "\n",
    "# seurat <- subset(x = seurat, subset = customclassif != \"Unknown\")\n",
    "\n",
    "# # Outputs final tSNE plot as pdf with clusters labeled\n",
    "# # Change name based on resolution\n",
    "# pdf(\"nov15-W7F/GFP_UMAP_markergenes_nov15.pdf\", width = 10, height = 10)\n",
    "# plot <- DimPlot(seurat, reduction = \"umap\", label = TRUE, label.size = 4, repel = TRUE, group.by = 'customclassif') \n",
    "# print(plot+ ggtitle(\"GFP UMAP\"))\n",
    "# dev.off()\n",
    "\n",
    "# # ---------------------------------------------------------------\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c3de41e-2f47-4323-8f0a-1db8ee76426c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
